File: src/main/java/com/jore/epoc/bo/orders/MarketingCampaignOrder.java
package com.jore.epoc.bo.orders;

import org.hibernate.annotations.CompositeType;

import com.jore.datatypes.money.Money;
import com.jore.epoc.bo.accounting.FinancialAccounting;
import com.jore.epoc.bo.message.MessageLevel;

import jakarta.persistence.AttributeOverride;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;

@Entity
public class MarketingCampaignOrder extends AbstractSimulationOrder {
    @AttributeOverride(name = "amount", column = @Column(name = "marketing_campaign_amount"))
    @AttributeOverride(name = "currency", column = @Column(name = "marketing_amount_currency"))
    @CompositeType(com.jore.datatypes.hibernate.MoneyCompositeUserType.class)
    private Money marketingCampaignAmount;

    @Override
    public void execute() {
        if (getCompany().getAccounting().checkFunds(marketingCampaignAmount)) {
            runMarketingCampaign();
            book(getExecutionMonth().atDay(1), "Marketing campaign", FinancialAccounting.SERVICES, FinancialAccounting.BANK, marketingCampaignAmount);
            addMessage(MessageLevel.INFORMATION, "MarketingCampaign", getExecutionMonth(), marketingCampaignAmount);
            setExecuted(true);
        } else {
            addMessage(MessageLevel.WARNING, "NoMarketingCampaignDueToFunding", getExecutionMonth(), marketingCampaignAmount, getCompany().getAccounting().getBankBalance());
        }
    }

    @Override
    public int getSortOrder() {
        return 8;
    }

    public void setAmount(Money marketingCampaignAmount) {
        this.marketingCampaignAmount = marketingCampaignAmount;
    }

    private void runMarketingCampaign() {
        Money pricePerMarketingCampaign = getCompany().getSimulation().getSettings().getPricePerMarketingCampaign();
        double increase = marketingCampaignAmount.divide(pricePerMarketingCampaign).doubleValue();
        increase /= 100;
        getCompany().setMarketingFactor(getCompany().getMarketingFactor() + increase);
    }
}

File: EpocServer/Sources/App/Models/Orders/MarketingCampaignOrder.swift
import Foundation
import Score

public final class MarketingCampaignOrder: SimulationOrder {
    public var id: Int64?
    public var executionMonth: YearMonth = .of(2020, 1)
    public var isExecuted: Bool = false
    public weak var company: Company!
    public var marketingCampaignAmount: Money = Money.of("CHF", 0)

    public var sortOrder: Int { 8 }
    public var type: String { "Marketing campaign" }

    public func getAmount() -> Money { marketingCampaignAmount }

    public func execute() {
        if company.accounting.checkFunds(marketingCampaignAmount, valueDate: executionMonth.atEndOfMonth()) {
            runMarketingCampaign()
            book(bookingDate: executionMonth.atDay(1), bookingText: "Marketing campaign",
                 debitAccount: FinancialAccounting.SERVICES, creditAccount: FinancialAccounting.BANK, amount: marketingCampaignAmount)
            addMessage(.information, "MarketingCampaign", executionMonth, marketingCampaignAmount)
            isExecuted = true
        } else {
            addMessage(.warning, "NoMarketingCampaignDueToFunding", executionMonth, marketingCampaignAmount, company.accounting.getBankBalance(executionMonth.atEndOfMonth()))
        }
    }

    private func runMarketingCampaign() {
        let pricePerCampaign = company.simulation.settings!.getPricePerMarketingCampaign()
        var increase = marketingCampaignAmount.divide(pricePerCampaign).doubleValue
        increase /= 100
        company.marketingFactor += increase
    }
}
